contributors/design-proposals/scheduling/cluster-scoped-resources.md

```diff
@@ -0,0 +1,88 @@
+
+#### Use Cases
+1. Software Licenses that can be shared by pods across the entire cluster
+
+2. IP Addresses:
+   Nodes in a cluster can be partitioned into multiple network scopes and each network scope can have a certain number of available IPs that can be assigned to pods running on those nodes. Hence, the nodes belonging to a network scope can collectively run as many pods as the IP capacity in that scope.
+
+3. Rack storage:
+   Locally attached shared storage in a rack, which is consumable by pods on nodes within a rack
+
+4. Network Bandwidth:
+   Network bandwidth shared by pods. Depending on the network topology of a cluster, pods on multiple nodes will be sharing network bandwidth with each other. In use cases where pods need guranteed network throughput, representing bandwidth as a cluster resource is essential for scheduling such pods.
+
+## Goals
+The goal is to define mechanisms to expose and consume cluster scoped resources
+
+## Design
+
+### ClusterResource type
+```
+// pkg/api/types.go:
+
+// ClusterResourceQuantity represents quantity of a ClusterResource
+type ClusterResourceQuantity struct {
+	Quantity resource.Quantity `json:"quantity"`
+	// NodeSelector is a label query over nodes which collectively provide resource Quantity
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+}
+
+// ClusterResource represents a resource which is available at a cluster level
+type ClusterResource struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+	// +optional
+	Status ClusterResourceStatus `json:"status,omitempty"`
+}
+
+type ClusterResourceStatus struct {
+	// Capacity represents the total quantity of ClusterResource
+	// +optional
+	Capacity []ClusterResourceQuantity `json:"capacity,omitempty"`
+	// Allocatable represents the quantity of ClusterResource that is available for scheduling
+	// +optional
+	Allocatable []ClusterResourceQuantity `json:"allocatable,omitempty"`
+}
+```
+`ClusterResourceStatus` captures the capacity and allocatable quantity for a `ClusteResource` in the form of `ClusterResourceQuantity`. `ClusterResourceQuantity` represents the quantity of a `ClusterResource` which is collectively consumable on nodes selected by `NodeSelector`.
+`NodeSelector` is a label query over the nodes, which collectively provide this `ClusterResource`. This field is optional, and if not specfied, it means that the `ClusterResource` is consumable across all nodes in the cluster.
+
+
+### Consuming ClusterResources
+
+ClusterResources are consumable by pods just like CPU and memory, by specifying it in the pod request. The scheduler should take care of the resource accounting for ClusterResources so that no more than the available amount is simultaneously allocated to Pods. The prefix used to identify a ClusterResource coule be 
+```
+pod.alpha.kubernetes.io/cluster-resource-
+```
+
+### Accounting in scheduler
```

Comment:
We have extended resources and several types of first class resources in the scheduler already. I think it would be possible to come up with a single presentation that covers all of these types. For example, I don't see much of a difference between a cluster resource and extended resource from scheduler's point of view. An extended resource with an additional "type" can present a cluster resource.

Reply:
The key difference between these two resources is the "scope".
Currently extended resources are exposed as a part of node status because of their nature of being tied to a node, while cluster scoped resources have to be represented outside the scope of a node. But we can surely have a comprehensive API that covers both. From the scheduler's point of view, it will need some additional logic to calculate and cache the available capacity of a cluster scoped resource across a set of nodes

Reply:
Yes, the "ResourceClass" that @jiayingz is working on is an effort in that direction to provide a comprehensive API to represent various types of resources, including cluster resources.

Reply:
Yes. As @bsalamat mentioned, we are working on a new Resource API proposal that aims to provide a comprehensive API for both node-level resources and cluster-level resources. Here is the current PR:
https://github.com/kubernetes/community/pull/782
It is still WIP and the current plan is to focus on node-level resources during the initial phase. But I think even the initial API should help solve some of the listed problems here. Please take a look and let us know if you see any missing pieces.